doc/decisions/DR016-One-to-many-result-paging.md

```diff
@@ -0,0 +1,522 @@
+
+#### Pros
+
+- Whole batch is paged at once, retaining batch-awareness in the
+  manager for subsequent pages.
+- Conceptually simpler API for both host and manager if only a single
+  page is active across the whole batch.
+
+### Pros (page token specific)
+
+- Per-element callback means it's trivial to skip processing elements of
+  the input batch whose pages are exhausted.
+- Retains existing design in line with OpenAssetIO design principles,
+  i.e. non-paged signature is simply the same function but ignoring
+  optional arguments.
+
+#### Cons
+
+- Wasted processing of empty lists by the host when pages are exhausted
+  i.e. when there are still pages remaining for some inputs but not
+  others.
+- Batch elements cannot enter an error state in isolation, since paging
+  is whole-batch.
+- Collaborators working on manager plugins raised this as being
+  particularly tricky to implement.
+
+#### Cons (page token specific)
+
+- Likely that the manager plugin must combine multiple back-end page
+  tokens into one, adding complexity to ensure they are maintained in
+  sync.
+
+#### Cons (cursor object specific)
+
+- Additional metadata in the object is limited to whole-batch data.
+
+### Batched paging option 3 - Per-element paging
+
+E.g. Page token
+
+```python
+# Callback taking index of batch element and selected page of results.
+SuccessCallback = Callable[[int, List[EntityReference]], None]
+
+def relatedEntities(
+    batch: List[EntityReference],
+    relationship,
+    successCb: SuccessCallback,
+    errorCb,
+    limit: -1,
+    pageTokens: List[AbstractPageToken]
+): -> List[AbstractPageToken]: ...
+```
+
+E.g. Cursor object
+
+```python
+class AbstractPage:
+    def get(self) -> List[EntityReferences]:
+        ...
+    def next(self, pageSize):
+        ...
+
+# Callback taking index of batch element and cursor object.
+SuccessCallback = Callable[[int, AbstractPage], None]
+
+def relatedEntities(
+    batch: List[EntityReference],
+    relationship,
+    successCb: SuccessCallback,
+    errorCb): ...
+```
+
+#### Pros
+
+- Manager is under no obligation to consider errors in other streams
+  when implementing query behaviour, as pagination method need not be
+  synchronized.
+
+#### Pros (cursor object specific)
+
+- Individual cursors can be incremented independently, potentially
+  reducing wasted queries.
+
+#### Cons
+
+- Batching is very difficult to optimize for the manager, especially
+  after the first page.
+- Batch-optimized queries have unpredictable performance profile for
+  the host, given the wide range of potential implementations.
+
+### Outcome
+
+Given the reported difficulty of crafting queries that are both batch
+and paging-aware, we opt to go with Option 1 and split the paging API(s)
+from the batched API(s).
+
+This then leaves only two remaining solutions: (non-batched) page token
+or cursor objects.
+
+## Considering manager implementations
+
+### Using Python's DB-API standard
+
+#### Batch signature
```

Comment:
[question] The preceding sentence rules out batching, be good to remove or move to the end?

Reply:
Batching isn't ruled out, instead it is split out from paging, so there's a separate sketch that considers batching only without paging.

Reply:
> ... we opt to go with Option 1 and split the paging API(s)
from the batched API(s).
>
> This then leaves only two remaining solutions: (non-batched) page token
or cursor objects.

Reads as excluded!

Reply:
Ah OK, I'll update that text.

Reply:
I thought that was good though - that were trying to limit the options in the DR?